### PR TITLE
Parse jsdoc `@param` type literals

### LINKED ISSUE #11597: Intellisense for object properties defined in multi-line JSDoc comments
Currently it doesn't look like the TypeScript parser recognizes properties for an object parameter if they are defined with separate `@param` declarations. Would it be possible to add support for this?

image

image

Thanks!

### PR DESCRIPTION
**Pull request by sandersn**
Now Typescript supports the creation of anonymous types using successive `@param` lines in JSDoc:

```js
/**
 * @param {object} o - has a string and a number
 * @param {string} o.s - the string
 * @param {number} o.n - the number
 */
function f(o) { return o.s.length + o.n; }
```

This is equivalent to the Typescript syntax `{ s: string, n: number }`, but it allows per-property documentation, even for types that only need to be used in one place. (`@typedef` can be used for reusable types.)

If the type of the initial `@param` is `{object[]}`, then the resulting type is an array of the specified anonymous type:

```js
/**
 * @param {Object[]} os - has a string and a number
 * @param {string} os[].s - the string
 * @param {number} os[].n - the number
 */
function f(os) { return os[0].s; }
```

Finally, nested anonymous types can be created by nesting the pattern:

```js
/**
 * @param {Object[]} os - has a string and a number
 * @param {string} os[].s - the string
 * @param {object} os[].nested - it's nested because of the object type
 * @param {number} os[].nested.length - it's a number
 */
function f(os) { return os[0].nested.length; }
```

Implementation notes:

1. I refactored JSDocParameterTag and JSDocPropertyTag to JSDocPropertyLikeTag and modified its parsing to be more succinct. These changes make the overall change easier to read but are not strictly required.
2. parseJSDocEntityName accepts postfix[] as in `os[].nested.length`, but it doesn't check that usages are correct. Such checking would be easy to add but tedious and low-value.
3. `@typedef` doesn't support nested `@property` tags, but does support `object[]` types. This is mostly a practical decision, backed up by the fact that usejsdoc.org doesn't document nested types for `@typedef`.

Fixes #11597

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -1553,6 +1553,10 @@ namespace ts {
 
     /** Does the opposite of `getJSDocParameterTags`: given a JSDoc parameter, finds the parameter corresponding to it. */
     export function getParameterFromJSDoc(node: JSDocParameterTag): ParameterDeclaration | undefined {
```
Comment (andy-ms) [inline src/compiler/utilities.ts]: See the use of this function -- at `x` in `@param obj.x` we should return the symbol for `x`. Add a find-all-references test.
```ts
/**
 * @param obj.[|x|]
 */
function f(obj) {
    obj.[|x|];
}
```

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -2163,13 +2164,16 @@ namespace ts {
...
+                case SyntaxKind.JSDocParameterTag:
+                    if (node.parent.kind !== SyntaxKind.JSDocTypeLiteral) {
+                        break;
+                    }
+                    // falls through
                 case SyntaxKind.JSDocPropertyTag:
-                    return declareSymbolAndAddToSymbolTable(node as JSDocPropertyTag,
-                        (node as JSDocPropertyTag).isBracketed || ((node as JSDocPropertyTag).typeExpression && (node as JSDocPropertyTag).typeExpression.type.kind === SyntaxKind.JSDocOptionalType) ?
+                    return declareSymbolAndAddToSymbolTable(node as JSDocPropertyLikeTag,
+                        (node as JSDocPropertyLikeTag).isBracketed || ((node as JSDocPropertyLikeTag).typeExpression && (node as JSDocPropertyLikeTag).typeExpression.type.kind === SyntaxKind.JSDocOptionalType) ?
```
Comment (andy-ms) [inline src/compiler/binder.ts]: This expression is too large.
```ts
const prop = node as JSDocPropertyLikeTag;
const flags = prop.isBracketed || prop.typeExpression && prop.typeExpression.type.kind === SyntaxKind.JSDocOptionalType
    ? SymbolFlags.Property | SymbolFlags.Optional
    : SymbolFlags.Property;
return declareSymbolAndAddToSymbolTable(prop, flags, SymbolFlags.PropertyExcludes);
```

Also, `typeExpression` is declared non-optional, but here we test for its presence.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -5074,20 +5074,9 @@ namespace ts {
...
-                    }
-                    else {
-                        type = getTypeFromTypeNode(declaration.typeExpression.type);
-                    }
-                }
-                else {
-                    declaration = getDeclarationOfKind<TypeAliasDeclaration>(symbol, SyntaxKind.TypeAliasDeclaration);
-                    type = getTypeFromTypeNode(declaration.type);
-                }
+                const declaration = getDeclarationOfKind<JSDocTypedefTag>(symbol, SyntaxKind.JSDocTypedefTag) ||
```
Comment (andy-ms) [inline src/compiler/checker.ts]: `getDeclarationOfKind` hides a loop, so this loops twice. Could add a version of that function that takes two kinds.

File: src/compiler/types.ts [hunk 4]
```diff
@@ -2125,38 +2126,33 @@ namespace ts {
...
-        kind: SyntaxKind.JSDocPropertyTag;
+        fullName?: EntityName;
         name: Identifier;
-        /** the parameter name, if provided *before* the type (TypeScript-style) */
-        preParameterName?: Identifier;
-        /** the parameter name, if provided *after* the type (JSDoc-standard) */
-        postParameterName?: Identifier;
         typeExpression: JSDocTypeExpression;
+        /** Whether the property name came before the type -- non-standard for JSDoc, but Typescript-like */
+        isParameterNameFirst: boolean;
```
Comment (andy-ms) [inline src/compiler/types.ts]: If this applies to more than just parameters, it should be renamed.

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -407,9 +410,15 @@ namespace ts {
             case SyntaxKind.JSDocComment:
                 return visitNodes(cbNode, cbNodes, (<JSDoc>node).tags);
             case SyntaxKind.JSDocParameterTag:
-                return visitNode(cbNode, (<JSDocParameterTag>node).preParameterName) ||
-                    visitNode(cbNode, (<JSDocParameterTag>node).typeExpression) ||
-                    visitNode(cbNode, (<JSDocParameterTag>node).postParameterName);
+            case SyntaxKind.JSDocPropertyTag:
+                if ((node as JSDocPropertyLikeTag).isParameterNameFirst) {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Maybe this should wait for a separate PR, but I would much prefer to see code like:
```ts
const { isParameterNameFirst, fullName, typeExpression } = node as JSDocPropertyLikeTag;
```
in all of these cases, instead of repeated casting.

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -419,15 +428,19 @@ namespace ts {
             case SyntaxKind.JSDocTemplateTag:
                 return visitNodes(cbNode, cbNodes, (<JSDocTemplateTag>node).typeParameters);
             case SyntaxKind.JSDocTypedefTag:
-                return visitNode(cbNode, (<JSDocTypedefTag>node).typeExpression) ||
-                    visitNode(cbNode, (<JSDocTypedefTag>node).fullName) ||
-                    visitNode(cbNode, (<JSDocTypedefTag>node).name) ||
-                    visitNode(cbNode, (<JSDocTypedefTag>node).jsDocTypeLiteral);
+                if ((node as JSDocTypedefTag).typeExpression &&
```
Comment (andy-ms) [inline src/compiler/parser.ts]: It looks like `name` is a pointer to a node within `fullName`. So it would be visited twice if we were recursively calling `forEachChild`.

File: src/compiler/parser.ts [hunk 7]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
...
+                        postName = fullName;
                     }
                     else {
-                        preName = name;
+                        preName = fullName;
                         typeExpression = tryParseTypeExpression();
                     }
 
-                    const result = shouldParseParamTag ?
+                    const result: JSDocPropertyLikeTag = target ?
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Since `target` is now an enum, use an explicit comparison instead of treating it like a boolean.

File: src/compiler/parser.ts [hunk 8]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
                         parseExpected(SyntaxKind.CloseBracketToken);
                     }
 
-                    return { name, isBracketed };
+                    return { fullName, isBracketed };
                 }
 
-                function parseParameterOrPropertyTag(atToken: AtToken, tagName: Identifier, shouldParseParamTag: boolean): JSDocPropertyTag | JSDocParameterTag {
+                function isObjectOrObjectArrayTypeReference(node: TypeNode): boolean {
+                    return node.kind === SyntaxKind.ObjectKeyword ||
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Use a `switch`.

File: src/compiler/parser.ts [hunk 9]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
...
+                function parseParameterOrPropertyTag(atToken: AtToken, tagName: Identifier, target: PropertyLikeParse): JSDocPropertyLikeTag {
                     let typeExpression = tryParseTypeExpression();
                     skipWhitespace();
 
-                    const { name, isBracketed } = parseBracketNameInPropertyAndParamTag();
+                    const { fullName, isBracketed } = parseBracketNameInPropertyAndParamTag();
                     skipWhitespace();
 
-                    let preName: Identifier, postName: Identifier;
+                    let preName: EntityName, postName: EntityName;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: It no longer makes sense to have separate `preName` and `postName` variables. Just create `isParameterNameFirst` directly.

File: src/compiler/parser.ts (hunk 5 above)
Comment (sandersn) [inline src/compiler/parser.ts]: I will make the change here if you are willing to review it! :sweat_smile: I'll put it in a separate commit at the end which might make it easier.

File: src/compiler/parser.ts [hunk 10]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
...
+                        }
+                        else {
+                            return false;
+                        }
+                    }
+                    return parent.text === name.text;
+                }
 
-                function tryParseChildTag(parentTag: JSDocTypeLiteral): boolean {
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse.Property): JSDocTypeTag | JSDocPropertyTag | false;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Since there are only 2 possible targets, maybe there should just be 2 different functions `parseChildParmeterTag` and `parseChildPropertyTag` that both use `parseChildParameterOrPropertyTagCommon`?

File: src/compiler/parser.ts (hunk 6 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Nice catch! I added `fullName` later and didn't notice that `name` needed to be removed.

File: src/compiler/parser.ts [hunk 11]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
...
+                    if (typeExpression && isObjectOrObjectArrayTypeReference(typeExpression.type)) {
+                        let child: JSDocPropertyLikeTag | false;
+                        let jsdocTypeLiteral: JSDocTypeLiteral;
+                        const start = scanner.getStartPos();
+                        while (child = tryParse(() => parseChildParameterOrPropertyTag(PropertyLikeParse.Parameter, fullName))) {
+                            if (!jsdocTypeLiteral) {
+                                jsdocTypeLiteral = <JSDocTypeLiteral>createNode(SyntaxKind.JSDocTypeLiteral, start);
+                                jsdocTypeLiteral.jsDocPropertyTags = [] as MutableNodeArray<JSDocPropertyTag>;
+                            }
+                            (jsdocTypeLiteral.jsDocPropertyTags as MutableNodeArray<JSDocPropertyTag>).push(child as JSDocPropertyTag);
```
Comment (andy-ms) [inline src/compiler/parser.ts]: You could just the create the mutable array locally and assign to `jsdocTypeLiteral`at the end. That way you avoid casting an already-created object to mutable.
This loop could probably be its own function `parseParameterTags` returning `NodeArray<JSDocPropertyTag> | undefined`.
Also, you don't appear to be setting the position on the array -- if you don't want to do that, just declare it as an `ReadonlyArray` and not a `NodeArray`.
Also, what's with the cast on `child`? You passed in `PropertyLikeParse.Parameter` so it's surprising to see an assertion that what you got back *isn't* a parameter.

File: src/compiler/parser.ts [hunk 12]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
...
+                        }
+                    }
                     result.atToken = atToken;
                     result.tagName = tagName;
-                    result.preParameterName = preName;
                     result.typeExpression = typeExpression;
-                    result.postParameterName = postName;
-                    result.name = postName || preName;
+                    if (typeExpression) {
+                        result.type = typeExpression.type;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Isn't this redundant? Why do we have both `typeExpression` and `type`?

File: src/compiler/parser.ts [hunk 13]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
...
+                    }
+                    result.fullName = postName || preName;
+                    result.name = ts.isIdentifier(result.fullName) ? result.fullName : result.fullName.right;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: This is also redundant, I would just make this a function of the node and not a property.

File: src/compiler/parser.ts [hunk 14]
```diff
@@ -6472,36 +6490,66 @@ namespace ts {
...
+                    result.isParameterNameFirst = postName ? false : !!preName;
                     result.isBracketed = isBracketed;
                     return finishNode(result);
+
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Nit: line before a closing bracket doesn't really accomplish anything -- nothing to separate that isn't already separated by the closing bracket.

File: src/compiler/parser.ts (hunk 5 above)
Comment (andy-ms) [inline src/compiler/parser.ts]: Please make a separate PR as that would be a huge diff.

File: src/compiler/parser.ts [hunk 15]
```diff
@@ -6565,68 +6613,44 @@ namespace ts {
...
+                        let jsdocTypeLiteral: JSDocTypeLiteral;
+                        let alreadyHasTypeTag = false;
+                        const start = scanner.getStartPos();
+                        while (child = tryParse(() => parseChildParameterOrPropertyTag(PropertyLikeParse.Property))) {
+                            if (!jsdocTypeLiteral) {
+                                jsdocTypeLiteral = <JSDocTypeLiteral>createNode(SyntaxKind.JSDocTypeLiteral, start);
+                            }
+                            if (child.kind === SyntaxKind.JSDocTypeTag) {
+                                if (alreadyHasTypeTag) {
+                                    break;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: What happens to `child` in this case? We just drop it with no syntax error?

File: src/compiler/parser.ts [hunk 16]
```diff
@@ -6565,68 +6613,44 @@ namespace ts {
...
+                                }
+                                else {
+                                    jsdocTypeLiteral.jsDocTypeTag = child;
+                                    alreadyHasTypeTag = true;
+                                }
+                            }
+                            else {
+                                if (!jsdocTypeLiteral.jsDocPropertyTags) {
+                                    jsdocTypeLiteral.jsDocPropertyTags = [] as MutableNodeArray<JSDocPropertyTag>;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: See earlier comment on casting to `MutableNodeArray`.

File: src/compiler/parser.ts [hunk 17]
```diff
@@ -6565,68 +6613,44 @@ namespace ts {
...
+                                }
+                                (jsdocTypeLiteral.jsDocPropertyTags as MutableNodeArray<JSDocPropertyTag>).push(child);
+                            }
                         }
-                        if (!typedefTag.jsDocTypeLiteral) {
-                            typedefTag.jsDocTypeLiteral = <JSDocTypeLiteral>typeExpression.type;
+                        if (jsdocTypeLiteral) {
+                            if (typeExpression && typeExpression.type.kind === SyntaxKind.ArrayType) {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: What are our performance guidelines on optionally vs unconditionally assigning to a property?

File: src/compiler/parser.ts [hunk 18]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
                     }
                 }
 
+                function textsEqual(parent: EntityName, name: EntityName): boolean {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Nit: would just name these variables `a` and `b` since they don't really depend on a parent-child context.

File: src/compiler/parser.ts [hunk 19]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
...
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse.Property): JSDocTypeTag | JSDocPropertyTag | false;
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse.Parameter, fullName: EntityName): JSDocPropertyTag | JSDocParameterTag | false;
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse, fullName?: EntityName): JSDocTypeTag | JSDocPropertyTag | JSDocParameterTag | false {
+                    let resumePos = scanner.getStartPos();
+                    let canParseTag = true;
+                    let seenAsterisk = false;
+                    while (token() !== SyntaxKind.EndOfFileToken) {
+                        nextJSDocToken();
+                        switch (token()) {
+                        case SyntaxKind.AtToken:
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Nit: indent switch body

File: src/compiler/parser.ts [hunk 20]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
...
+                            if (seenAsterisk) {
+                                canParseTag = false;
+                            }
+                            seenAsterisk = true;
+                            break;
+                        case SyntaxKind.Identifier:
+                            canParseTag = false;
+                            break;
+                        case SyntaxKind.EndOfFileToken:
+                            break;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Nit: I would do `loop: while (true) {` and `break loop;` here.

File: src/compiler/parser.ts [hunk 21]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
...
+                    let seenAsterisk = false;
+                    while (token() !== SyntaxKind.EndOfFileToken) {
+                        nextJSDocToken();
+                        switch (token()) {
+                        case SyntaxKind.AtToken:
+                            if (canParseTag) {
+                                const child = tryParseChildTag(target);
+                                if (child && child.kind === SyntaxKind.JSDocParameterTag &&
+                                    (ts.isIdentifier(child.fullName) || !textsEqual(fullName, child.fullName.left))) {
+                                    break;
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Should this be `break loop`? It looks like the way it is currently it goes all the way to the end of the file, when we could stop the moment we see something that isn't an `@param` tag with the correct name.

File: src/compiler/parser.ts [hunk 22]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
...
+                            seenAsterisk = true;
+                            break;
+                        case SyntaxKind.Identifier:
+                            canParseTag = false;
+                            break;
+                        case SyntaxKind.EndOfFileToken:
+                            break;
+                        }
+                    }
+                    scanner.setTextPos(resumePos);
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Do we have an established pattern for re-setting the scanner? What is `scanner` has some other state that this doesn't reset?

File: src/compiler/parser.ts [hunk 23]
```diff
@@ -6647,8 +6671,61 @@ namespace ts {
...
+                }
+
+                function tryParseChildTag(target: PropertyLikeParse, alreadyHasTypeTag?: boolean): JSDocTypeTag | JSDocPropertyTag | JSDocParameterTag | false {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: This is only called in one place, and `alreadyHasTypeTag` is never provided.

File: src/compiler/parser.ts [hunk 24]
```diff
@@ -6659,27 +6736,16 @@ namespace ts {
...
-                                if (!parentTag.jsDocPropertyTags) {
-                                    parentTag.jsDocPropertyTags = <MutableNodeArray<JSDocPropertyTag>>[];
-                                }
-                                (parentTag.jsDocPropertyTags as MutableNodeArray<JSDocPropertyTag>).push(propertyTag);
-                                return true;
-                            }
-                            // Error parsing property tag
-                            return false;
+                            return target === PropertyLikeParse.Property && parseParameterOrPropertyTag(atToken, tagName, target);
+                        case "arg":
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Is the information about what tag name was used available in the AST? That might be useful for linting.

File: src/compiler/parser.ts [hunk 25]
```diff
@@ -6728,6 +6794,26 @@ namespace ts {
                     return currentToken = scanner.scanJSDocToken();
                 }
 
+                function parseJSDocEntityName(createIfMissing = false): EntityName {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: This is only called in one place, and `createIfMissing` is always true.

Comment (andy-ms) [inline src/compiler/parser.ts]: This could use:

```ts
function createQualifiedName(entity: EntityName, name: Identifier): QualifiedName {
    const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
    node.left = entity;
    node.right = name;
    return finishNode(node);
}
```

And you can re-use `createQualifiedName` in `parseEntityName`. (`parseJSDocEntityName` should probably be near `parseEntityName` since they are similar.)

File: src/services/classifier.ts [hunk 26]
```diff
@@ -767,10 +767,10 @@ namespace ts {
                     pos = tag.typeExpression.end;
                 }
 
-                if (tag.postParameterName) {
-                    pushCommentRange(pos, tag.postParameterName.pos - pos);
-                    pushClassification(tag.postParameterName.pos, tag.postParameterName.end - tag.postParameterName.pos, ClassificationType.parameterName);
-                    pos = tag.postParameterName.end;
+                if (!tag.isParameterNameFirst) {
+                    pushCommentRange(pos, tag.name.pos - pos);
```
Comment (andy-ms) [inline src/services/classifier.ts]: These two `if` blocks are the same, so use this helper:
```ts
function pushName(name: Identifier) {
    pushCommentRange(pos, name.pos - pos);
    pushClassification(name.pos, name.end - name.pos, ClassificationType.parameterName);
    pos = name.end;
}
```

File: src/services/jsDoc.ts [hunk 27]
```diff
@@ -120,7 +120,7 @@ namespace ts.JsDoc {
     }
 
     export function getJSDocParameterNameCompletions(tag: JSDocParameterTag): CompletionEntry[] {
-        const nameThusFar = unescapeLeadingUnderscores(tag.name.text);
+        const nameThusFar = isIdentifier(tag.fullName) ? unescapeLeadingUnderscores(tag.name.text) : undefined;
```
Comment (andy-ms) [inline src/services/jsDoc.ts]: I think we should bail out immediately if it's not an identifier, instead of continuing with an undefined `nameThusFar`.

File: src/compiler/utilities.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/utilities.ts]: Added.

File: src/compiler/checker.ts (hunk 3 above)
Comment (sandersn) [inline src/compiler/checker.ts]: I'll just use `findDeclaration` which takes a predicate.

**Comment by sandersn** [comment]
@weswigham 
1. The parser defines its own `isIdentifier` which operates while parsing an identifier, unlike the normal `ts.isIdentifier`.
2. I kind of like the idea of having different syntax kinds for (1) normal `@param` (2) type literal `@param` (3) non-standard `@param`, but I'm not sure it's worth the effort. I'll think about it and maybe include it in this PR and maybe not.

File: src/compiler/checker.ts (hunk 3 above)
Comment (andy-ms) [inline src/compiler/checker.ts]: I was going to delete that function in #17380 since it's equivalent to just `find(symbol.declarations, ...)`.

File: src/compiler/parser.ts (hunk 11 above)
Comment (sandersn) [inline src/compiler/parser.ts]: I don't know a correct position to set on the child array because I don't know what would use it. I don't think I can make it a ReadonlyArray though, because forEachChild expects to iterate over a NodeArray. For now I'll still create a NodeArray with no position.

Comment (andy-ms) [inline src/compiler/parser.ts]: You could just use a for loop instead of calling `cbNodes`, which does require a NodeArray, meaning users might be expecting it to have a position.

Comment (sandersn) [inline src/compiler/parser.ts]: Sure. Done.

File: src/compiler/parser.ts (hunk 17 above)
Comment (sandersn) [inline src/compiler/parser.ts]: I thought the standard for booleans was to declare them as `b?: boolean` and only assign them when needed. What difference in performance could arise? Conditional assignment saves space (but how much I don't know) versus unconditional assignment, which could be better for [de]optimisation. @rbuckton do you have an opinion?

File: src/compiler/parser.ts (hunk 18 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Done.

File: src/compiler/parser.ts (hunk 19 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Fixed

File: src/compiler/parser.ts (hunk 10 above)
Comment (sandersn) [inline src/compiler/parser.ts]: I think a more likely refactoring would be to pass 1 of 2 different `tryParseChildTag` functions instead of the `target` enum. But the `target` enum pattern is also used with `parseParameterOrPropertyTag`, and it's as easy to read and (probably?) as fast to run, so I think I'll keep it this way.

File: src/compiler/parser.ts (hunk 21 above)
Comment (sandersn) [inline src/compiler/parser.ts]: This function is always called from in a speculative context, so I can actually just return false.

File: src/compiler/parser.ts (hunk 20 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Same here, I can just return false.

File: src/compiler/parser.ts (hunk 22 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Yep, this is left over from the old `scanChildTags` that faked its own speculative parsing. Removed.

File: src/compiler/parser.ts (hunk 23 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Oops, I moved that up a couple of levels and forgot to remove it here.

File: src/compiler/parser.ts (hunk 24 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Yes, `JSDocTag.tagName: Identifier`, so it looks like it's preserved on all JSDocTags.

A more useful lint rule would be to look for mismatched `typedef/param and param/property` nestings, since the compiler currently ignores them:

```js
/**
 * @typedef {object} x
 * @param {string} x.a - WRONG! (but ignored by the compiler)
 *
 * @param {object} y
 * @property {string} y.a - WRONG! (but ignored by the compiler)
 */
```

File: src/compiler/parser.ts (hunk 25 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Done. I thought about merging `parseEntityName` and `parseJSDocEntityName`, but they both worry about ad-hoc jsdoc-related things, plus they call into different identifier parsers which do the same.

File: src/services/jsDoc.ts (hunk 27 above)
Comment (sandersn) [inline src/services/jsDoc.ts]: Done. I returned emptyArray instead of undefined. I'm not sure whether that's correct.

File: src/compiler/parser.ts (hunk 8 above)
Comment (andy-ms) [inline src/compiler/parser.ts]: `isTypeReferenceNode` just tests the syntax kind, so just make that a switch case and have `default: return false;`.

File: src/compiler/parser.ts [hunk 28]
```diff
@@ -6655,8 +6692,58 @@ namespace ts {
...
+                }
+
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse.Property): JSDocTypeTag | JSDocPropertyTag | false;
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse.Parameter, name: EntityName): JSDocParameterTag | false;
+                function parseChildParameterOrPropertyTag(target: PropertyLikeParse, name?: EntityName): JSDocTypeTag | JSDocPropertyTag | JSDocParameterTag | false {
+                    let canParseTag = true;
+                    let seenAsterisk = false;
+                    while (true) {
+                        nextJSDocToken();
+                        switch (token()) {
```
Comment (andy-ms) [inline src/compiler/parser.ts]: Can we `return false;` in the `default` case? It looks like if we keep encountering some other kind of token, we'll go all the way to the end of the file.

File: src/services/classifier.ts (hunk 26 above)
Comment (andy-ms) [inline src/services/classifier.ts]: Don't know why this shows as outdated, it still seems relevant.

File: src/services/jsDoc.ts (hunk 27 above)
Comment (andy-ms) [inline src/services/jsDoc.ts]: Should be fine.

File: src/compiler/parser.ts (hunk 28 above)
Comment (sandersn) [inline src/compiler/parser.ts]: Not really; we hit the default case a lot for punctuation and identifiers that we want to skip as part of the tag's description. 

Note that it's the end of the comment, not the end of the whole file. And we return false as soon as `tryParseChildTag` encounters any tag that is not `param`, `type` or `property`. So runaway parsing shouldn't be a big problem.

Also note: Comments aren't handled by this PR but once they are, `default` will probably look like `parseJSDocCommentWorker`, where it just pushes the token into a comment array.

**Comment by sandersn** [comment]
I think this is ready to go unless you have more comments.

File: src/compiler/parser.ts (hunk 15 above)
Comment (andy-ms) [inline src/compiler/parser.ts]: I think this comment is still unaddressed?

Comment (sandersn) [inline src/compiler/parser.ts]: Yes. It happens in this case:

```ts
/**
 * @typedef Name
 * @type {string}
 * @type {Oops} - this is dropped and we stop parsing the @typedef
 */
 ```

This would be a grammar error if we did more thorough checking of jsdoc structure in the checker (along with a lot of other incorrect-but-allowed jsdoc). We don't, which is OK relative to what we used to provide. We should probably improve in the future if jsdoc becomes the source of types for significant, typed codebases.

**Comment by sandersn** [comment]
Addressed the comment I missed previously.

File: src/compiler/parser.ts (hunk 15 above)
Comment (andy-ms) [inline src/compiler/parser.ts]: OK, added #17442.
